cmd/common/image_functions.go

```diff
@@ -150,43 +152,147 @@ func GetLastTagFromResponse(resultTags *acr.RepositoryTagsType) string {
+				if manifest.MediaType != nil {
+					// If the manifest is not a list type, we can skip searching for its children
+					if *manifest.MediaType != v1.MediaTypeImageIndex && *manifest.MediaType != mediaTypeDockerManifestList {
+						continue
+					}
+				}
+				group.SubmitErr(func() error {
+					// For tagged indexes/lists, we need to get dependencies and add them to ignore list
+					// We don't need to check deletability since it's tagged (not deletable), just get dependencies
+					manifestBytes, err := acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+					if err != nil {
+						errParsed := autorest.DetailedError{}
+						if errors.As(err, &errParsed) && errParsed.StatusCode == http.StatusNotFound {
+							// If manifest not found, skip it
+							return nil
+						}
+						return err
 					}
-				} else {
-					if *manifest.MediaType != v1.MediaTypeImageManifest {
-						candidates = append(candidates, manifest)
+
+					dependentManifests, err := extractSubmanifestsFromBytes(manifestBytes)
+					if err != nil {
+						return err
+					}
+
+					if len(dependentManifests) > 0 {
+						return addDependentManifestsToIgnoreList(ctx, dependentManifests, acrClient, repoName, &ignoreList)
+					}
+					return nil
+				})
+				continue // We can skip the rest since the index is tagged and we are going to find its children
+			}
+
+			// _____MANIFEST IS UNTAGGED BUT MAY BE PROTECTED_____
+			// TODO: #468 I am a little unclear as to why this was ever an option but respecting it for now. Its not used by the purge scenarios only for
+			// the annotate command.
+			if preserveAllOCIManifests {
+				if *manifest.MediaType != v1.MediaTypeImageManifest {
+					// Add the manifest to the candidates list
+					if _, ok := candidates[*manifest.Digest]; !ok {
+						candidates[*manifest.Digest] = manifest
 					}
 				}
+				continue
+			}
+
+			// ______MANIFEST IS UNTAGGED BUT MAY STILL BE A REFERRER_____
+			// If the manifest is a referrer type we want to preserve it as ACR does cleanup on the server side when a parent manifest is deleted
+
+			// Schedule a goroutine to check if the manifest is okay to delete, it will be added to the
+			// candidates list anyway but if it is not okay to delete we will add it to the ignoreList
+
+			// We only need to do this check if we are looking at an oci index or oci manifest
+			group.SubmitErr(func() error {
+				canDelete, dependentManifests, err := checkManifestDeletabilityAndGetDependencies(ctx, manifest, acrClient, repoName)
+				if err != nil {
+					return err
+				}
+				if canDelete {
+					// Manifest is okay to delete
+					return nil
+				}
+
+				// If the manifest has dependencies (is an index), add them to the ignore list
+				if len(dependentManifests) > 0 {
+					return addDependentManifestsToIgnoreList(ctx, dependentManifests, acrClient, repoName, &ignoreList)
+				}
+
+				ignoreList.LoadOrStore(*manifest.Digest, struct{}{})
```

Comment:
is this thread safe?

Reply:
It is yep. The sync.Map is meant for concurrent access